src/backend/var_array.rs

```diff
@@ -0,0 +1,55 @@
+/// A C-style variable length array implemented as one allocation with
+/// a prepended header.
+use libpulse_sys::{pa_xrealloc};
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct VarArray<T>
+{
+    len: u32,
+    data: [T;0],
+}
+
+impl<T> VarArray<T>
+{
+    pub fn len(&self) -> usize {
+        self.len as usize
+    }
+
+    unsafe fn _realloc(ptr: Option<Box<Self>>, count: usize) -> Box<Self> {
+        use std::mem::{size_of, transmute};
+
+        let size = size_of::<Self>() + count * size_of::<T>();
+        let raw_ptr = match ptr {
+            Some(box_ptr) => Box::into_raw(box_ptr) as *mut u8,
+            None => 0 as *mut u8
+        };
+        let mem = pa_xrealloc(raw_ptr as *mut _, size);
```

Comment:
Because the C version has access to malloc/free, etc. It was convenient to use PA's zeroing wrappers for lib realloc, et al., because those are exposed, especially after removing the dependency upon libc.